Sources/SwiftKafka/KafkaBackPressurePollingSystem.swift

```diff
@@ -0,0 +1,274 @@
+
+            switch action {
+            case .poll:
+                self.client?.withKafkaHandlePointer { handle in
+                    rd_kafka_poll(handle, 0)
+                }
+                try? await Task.sleep(nanoseconds: pollIntervalNanos)
+            case .suspendPollLoop:
+                await withTaskCancellationHandler {
+                    await withCheckedContinuation { continuation in
+                        self.stateMachine.eventTriggered(.suspendLoop(continuation))
+                    }
+                } onCancel: {
+                    self.stateMachine.eventTriggered(.loopSuspensionCancelled)
+                }
+            case .shutdownPollLoop:
+                self.stateMachine.eventTriggered(.shutdown)
+                break actionLoop
+            }
+        }
+    }
+
+    // TODO: Race between poll loop and delivery report callback?
+    /// The delivery report callback function that handles acknowledged messages.
+    private(set) lazy var deliveryReportCallback: (UnsafePointer<rd_kafka_message_t>?) -> Void = { messagePointer in
+        guard let messagePointer = messagePointer else {
+            self.logger.error("Could not resolve acknowledged message")
+            return
+        }
+
+        let messageID = UInt(bitPattern: messagePointer.pointee._private)
+
+        let messageResult: Element
+        do {
+            let message = try KafkaAcknowledgedMessage(messagePointer: messagePointer, id: messageID)
+            messageResult = .success(message)
+        } catch {
+            guard let error = error as? KafkaAcknowledgedMessageError else {
+                fatalError("Caught error that is not of type \(KafkaAcknowledgedMessageError.self)")
+            }
+            messageResult = .failure(error)
+        }
+
+        guard let yieldResult = self.sequenceSource?.yield(messageResult) else {
+            fatalError("\(#function) requires the sequenceSource variable to be non-nil")
+        }
+
+        switch yieldResult {
+        case .produceMore:
+            self.stateMachine.eventTriggered(.produceMore)
+        case .stopProducing:
+            self.stateMachine.eventTriggered(.stopProducing)
+        case .dropped:
+            self.stateMachine.eventTriggered(.shutdown)
+        }
+
+        // The messagePointer is automatically destroyed by librdkafka
+        // For safety reasons, we only use it inside of this closure
+    }
+}
+
+extension KafkaBackPressurePollingSystem: NIOAsyncSequenceProducerDelegate {
+    func produceMore() {
+        self.stateMachine.eventTriggered(.produceMore)
+    }
+
+    func didTerminate() {
+        self.stateMachine.eventTriggered(.shutdown)
+    }
+}
+
+extension KafkaBackPressurePollingSystem {
+    // TODO: test state machine
+
+    /// The state machine used by the ``KafkaBackPressurePollingSystem``.
+    struct StateMachine {
+        /// The possible actions for the poll loop.
+        enum PollLoopAction {
```

Comment:
Can we move the actions right above the method that returns them. This makes it easier to understand what is going on.